### PR TITLE
Remove text align control for paragraph, heading, and button in contentOnly editing mode

### PR DESCRIPTION
**Pull request by talldan**
## What?
See #53705

In `contentOnly` block editing mode, only controls for 'content' attributes should be visible.

For paragraph, heading and button, the text align control was visible in the toolbar, but this isn't a content attribute.

Also removes the heading level tool in contentOnly mode.

## Why?
This causes an issue for pattern overrides. For those editing a paragraph in a pattern instance, it looks like the text align can be changed, but this change is never saved since the attribute isn't supported as an override.

## How?
Use `useBlockEditingMode` to get the editing mode, and only show these controls when the mode is `default`.

## Testing Instructions
1. Follow the testing instructions here to create a pattern with overrides - https://github.com/WordPress/gutenberg/issues/53705#issuecomment-1882305331
2. Insert the created pattern in a post
3. Try editing a paragraph or button, you should not see the text align control on the toolbar-->

### Before
Screenshot 2024-01-17 at 2 28 03 pm

### After
Screenshot 2024-01-17 at 2 27 22 pm

### CODE REVIEW

**Review by kevin940726 — APPROVED** [review]
Works as advertised!

What do you think of the same control for the heading block though?

image

**Comment by talldan** [comment]
> What do you think of the same control for the heading block though?

Good point. I'll push a commit that does the same there.

edit: the heading level also needs to be removed.

**Comment by talldan** [comment]
Done 👍

**Review by kevin940726 — APPROVED** [review]
Nice! 💯

File: packages/block-library/src/heading/edit.js [hunk 1]
```diff
@@ -90,20 +92,22 @@ function HeadingEdit( {
...
-				/>
-			</BlockControls>
+			{ blockEditingMode === 'default' && (
+				<BlockControls group="block">
+					<HeadingLevelDropdown
+						value={ level }
+						onChange={ ( newLevel ) =>
+							setAttributes( { level: newLevel } )
+						}
+					/>
```
Comment (fabiankaegy) [inline packages/block-library/src/heading/edit.js:102]: I disagree with the removal of this. Heading level is not a visual setting but rater a semantic one. Depending on where a pattern is used adjusting the heading level is required to make sure content follows good semantics and is as accessible as possible.

Comment (talldan) [reply packages/block-library/src/heading/edit.js:102]: True, though we'd need to explore making it work as an attribute that supports block binding, as it wasn't working. 

I guess there is still a question of whether it should be retained as viewable in contentOnly mode outside of patterns.

I'll add it to the tracking issue to explore.

Comment (fabiankaegy) [reply packages/block-library/src/heading/edit.js:102]: I would consider this a breaking change for how content only mode currently functions. 

I get there are implications for the new block connections and overrides.

But this is making the existing content only locking mode less usable.

Comment (talldan) [reply packages/block-library/src/heading/edit.js:102]: It was really a bug, as that attribute isn't marked `role: 'content'`, so shouldn't have been visible in 'contentOnly' mode from the start. The same with text align.

I understand the point that it limits important functionality though. I think it's one of those attributes that falls between the cracks in not being content, but being important for the structure of content.

I think there's some suffering from 'contentOnly' mode not being properly thought out, and personally I don't think the `role: 'content'` attributes were a good idea from the start. It's definitely an area that needs more work, and I think post 6.5 pattern overrides may lean on it less in favour of something a bit more expressive.

File: packages/block-library/src/heading/edit.native.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
+/**
```
Comment (ellatrix) [inline packages/block-library/src/heading/edit.native.js:1]: Hm, why was this duplicated?

File: packages/block-library/src/button/edit.js [hunk 3]
```diff
@@ -277,12 +279,14 @@ function ButtonEdit( props ) {
...
 			</div>
 			<BlockControls group="block">
-				<AlignmentControl
-					value={ textAlign }
-					onChange={ ( nextAlign ) => {
-						setAttributes( { textAlign: nextAlign } );
-					} }
-				/>
+				{ blockEditingMode === 'default' && (
+					<AlignmentControl
```
Comment (ellatrix) [inline packages/block-library/src/button/edit.js:283]: It sounds to me like the `blockEditingMode` should be part of the alignment control component.

File: packages/block-library/src/paragraph/edit.js [hunk 4]
```diff
@@ -108,28 +109,31 @@ function ParagraphBlock( {
 		} ),
 		style: { direction },
 	} );
+	const blockEditingMode = useBlockEditingMode();
```
Comment (ellatrix) [inline packages/block-library/src/paragraph/edit.js:112]: This is adding a store subscription for every paragraph block and is thus causing a performance regression.

File: packages/block-library/src/paragraph/edit.js [hunk 5]
```diff
@@ -108,28 +109,31 @@ function ParagraphBlock( {
...
+						onChange={ ( newAlign ) =>
+							setAttributes( {
+								align: newAlign,
+								dropCap: hasDropCapDisabled( newAlign )
+									? false
+									: dropCap,
+							} )
+						}
+					/>
+					<ParagraphRTLControl
```
Comment (ellatrix) [inline packages/block-library/src/paragraph/edit.js:129]: Why was this moved in the condition?

File: packages/block-library/src/heading/edit.native.js (hunk 2 above)
Comment (talldan) [reply packages/block-library/src/heading/edit.native.js:1]: Native code doesn't support `useBlockEditingMode`

File: packages/block-library/src/paragraph/edit.js (hunk 5 above)
Comment (talldan) [reply packages/block-library/src/paragraph/edit.js:129]: Why not? Not sure it's something that should appear for `contentOnly` or `disabled` blocks.

File: packages/block-library/src/paragraph/edit.js (hunk 4 above)
Comment (talldan) [reply packages/block-library/src/paragraph/edit.js:112]: Thanks for catching it. There's a few blocks that use `useBlockEditingMode` like this, so that's why I didn't think it would be much of an issue, but I should have checked the performance results.

File: packages/block-library/src/button/edit.js (hunk 3 above)
Comment (talldan) [reply packages/block-library/src/button/edit.js:283]: That's a good option. There are a few controls that have this kind of implementation.

It's a little hard to avoid this issue - https://github.com/WordPress/gutenberg/pull/57941#discussion_r1456713064.

I'll look into if there's an alternative. 👍

File: packages/block-library/src/paragraph/edit.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-library/src/paragraph/edit.js:129]: It's a language thing no? That's content?
